File: 0056. Merge Intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        ans = []
        for i, interval in enumerate(
            sorted(
                intervals,
                key=lambda interval: interval[0]
            )
        ):
            if i == 0:
                ans.append(interval)
                continue
            if ans[-1][1] < interval[0]:
                ans.append(interval)
                continue
            if interval[1] > ans[-1][1]:
                ans[-1][1] = interval[1]
        return ans

File: 0056. Merge Intervals.ts
// O(n log n) time and O(n) space complexity
function merge(intervals: number[][]): number[][] {
    // sort by interval starts
    const sortedIntervals: number[][] = intervals.sort((a, b) => a[0] - b[0]);

    // check for mergable intervals
    const mergedIntervals: number[][] = [];
    let min: number = intervals[0][0];
    let max: number = intervals[0][1];
    sortedIntervals.forEach((interval: number[]) => {
        // interval overlaps
        if (interval[0] <= max) {
            if (interval[1] > max) {
                max = interval[1];
            };
            return;
        };

        mergedIntervals.push([min, max]);
        min = interval[0];
        max = interval[1];
    });
    mergedIntervals.push([min, max]);

    return mergedIntervals;
};
